server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -241,14 +242,19 @@ private static NodeBalanceStats createFrom(
                 }
             }
 
+            double nodeWeight = desiredBalance == null || desiredBalance.weightsPerNode().isEmpty()
+                ? 0
```

Comment:
Both 0 and negatives are possible valid values. So I'd suggest we use `Double` for the field type and `readOptionalDouble/writeOptionalDouble` for read and write and skip when it is `null` in xcontent similar to how `writeLoadForecast` is [handled](https://github.com/elastic/elasticsearch/blob/b090ccd5c74fe2e38d2387c762c2d1e62e15dee0/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L1686) in IndexMetadata.